File: external/ghidra-server/scripts/ExportAnalysisJSON.py
# ExportAnalysisJSON.py - Export Ghidra analysis to JSON
# @category Analysis
# @author REVENG Team

import json
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

def export_analysis(output_file):
    """Export complete Ghidra analysis to JSON"""

    # Initialize decompiler
    decompiler = DecompInterface()
    decompiler.openProgram(currentProgram)

    # Get program info
    program = currentProgram
    listing = program.getListing()
    memory = program.getMemory()

    result = {
        "status": "success",
        "program_name": program.getName(),
        "language": str(program.getLanguageID()),
        "compiler": str(program.getCompilerSpec().getCompilerSpecID()),
        "functions": [],
        "strings": [],
        "imports": [],
        "exports": [],
        "xrefs": []
    }

    # Extract functions
    function_manager = program.getFunctionManager()
    functions = function_manager.getFunctions(True)  # True = forward iteration

    for func in functions:
        func_data = {
            "name": func.getName(),
            "entry_point": str(func.getEntryPoint()),
            "address": str(func.getEntryPoint()),
            "body": [],
            "source": "",
            "decompiled": "",
            "signature": str(func.getSignature()),
            "calling_convention": str(func.getCallingConventionName())
        }

        # Get assembly instructions
        instructions = listing.getInstructions(func.getBody(), True)
        for instr in instructions:
            func_data["body"].append({
                "address": str(instr.getAddress()),
                "mnemonic": str(instr.getMnemonicString()),
                "operands": str(instr.getDefaultOperandRepresentation(0)) if instr.getNumOperands() > 0 else ""
            })

        # Decompile function
        try:
            decompile_results = decompiler.decompileFunction(func, 30, ConsoleTaskMonitor())
            if decompile_results and decompile_results.decompileCompleted():
                decompiled_function = decompile_results.getDecompiledFunction()
                source = str(decompiled_function.getC()) if decompiled_function else ""
                func_data["source"] = source
                func_data["decompiled"] = source
            elif decompile_results:
                # Decompilation started but didn't complete - log error message
                error_msg = str(decompile_results.getErrorMessage()) if decompile_results.getErrorMessage() else "Unknown decompilation error"
                func_data["decompile_error"] = error_msg
        except Exception as e:
            # Decompilation exception
            func_data["decompile_error"] = str(e)

        result["functions"].append(func_data)

    result["functions"] = sorted(
        result["functions"],
        key=lambda func_data: len(func_data.get("source", "")),
        reverse=True,
    )

    # Extract strings
    defined_data = listing.getDefinedData(True)
    for data in defined_data:
        if data.hasStringValue():
            result["strings"].append({
                "address": str(data.getAddress()),
                "value": str(data.getValue()),
                "length": data.getLength()
            })

    # Extract imports
    external_manager = program.getExternalManager()
    ext_names = external_manager.getExternalLibraryNames()
    for lib_name in ext_names:
        externals = external_manager.getExternalLocations(lib_name)
        while externals.hasNext():
            ext_loc = externals.next()
            result["imports"].append({
                "name": ext_loc.getLabel(),
                "library": lib_name,
                "address": str(ext_loc.getExternalSpaceAddress()) if ext_loc.getExternalSpaceAddress() else "external"
            })

    # Extract exports - skip for now to avoid API issues
    # TODO: Fix exports extraction
    pass

    # Write JSON output
    with open(output_file, 'w') as f:
        json.dump(result, f, indent=2)

    print("Analysis exported to: " + output_file)
    print("Functions found: " + str(len(result["functions"])))
    print("Strings found: " + str(len(result["strings"])))
    print("Imports found: " + str(len(result["imports"])))

# Main execution
if __name__ == "__main__":
    # Get output file from script arguments
    args = getScriptArgs()
    if args and len(args) > 0:
        output_file = args[0]
    else:
        output_file = "/tmp/analysis.json"

    export_analysis(output_file)

File: external/ghidra-server/scripts/ExportAnalysisJSON.java
/* ###
 * Export Ghidra analysis to JSON.
 */
//@category Analysis

import java.io.FileWriter;
import java.io.Writer;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.listing.Program;
import ghidra.program.model.symbol.ExternalLocation;
import ghidra.program.model.symbol.ExternalLocationIterator;
import ghidra.program.model.symbol.ExternalManager;

public class ExportAnalysisJSON extends GhidraScript {

	@Override
	public void run() throws Exception {
		String[] args = getScriptArgs();
		String outputFile = (args != null && args.length > 0) ? args[0] : "analysis.json";

		Map<String, Object> result = exportAnalysis();
		Gson gson = new GsonBuilder().setPrettyPrinting().create();

		try (Writer writer = new FileWriter(outputFile)) {
			gson.toJson(result, writer);
		}

		println("Analysis exported to: " + outputFile);
		println("Functions found: " + ((List<?>) result.get("functions")).size());
		println("Strings found: " + ((List<?>) result.get("strings")).size());
		println("Imports found: " + ((List<?>) result.get("imports")).size());
	}

	private Map<String, Object> exportAnalysis() throws Exception {
		Program program = currentProgram;
		Listing listing = program.getListing();
		DecompInterface decompiler = new DecompInterface();
		decompiler.openProgram(program);

		Map<String, Object> result = new LinkedHashMap<>();
		result.put("status", "success");
		result.put("program_name", program.getName());
		result.put("language", program.getLanguageID().toString());
		result.put("compiler", program.getCompilerSpec().getCompilerSpecID().toString());

		List<Map<String, Object>> functions = new ArrayList<>();
		List<Map<String, Object>> strings = new ArrayList<>();
		List<Map<String, Object>> imports = new ArrayList<>();
		result.put("functions", functions);
		result.put("strings", strings);
		result.put("imports", imports);
		result.put("exports", new ArrayList<Map<String, Object>>());
		result.put("xrefs", new ArrayList<Map<String, Object>>());

		try {
			FunctionIterator functionIterator = program.getFunctionManager().getFunctions(true);
			while (functionIterator.hasNext() && !monitor.isCancelled()) {
				Function function = functionIterator.next();
				Map<String, Object> functionData = new LinkedHashMap<>();
				String entryPoint = function.getEntryPoint().toString();

				functionData.put("name", function.getName());
				functionData.put("entry_point", entryPoint);
				functionData.put("address", entryPoint);
				functionData.put("signature", function.getSignature().toString());
				functionData.put("calling_convention", String.valueOf(function.getCallingConventionName()));
				functionData.put("source", "");
				functionData.put("decompiled", "");

				List<Map<String, Object>> body = new ArrayList<>();
				functionData.put("body", body);
				InstructionIterator instructions = listing.getInstructions(function.getBody(), true);
				while (instructions.hasNext() && !monitor.isCancelled()) {
					Instruction instruction = instructions.next();
					Map<String, Object> instructionData = new LinkedHashMap<>();
					instructionData.put("address", instruction.getAddress().toString());
					instructionData.put("mnemonic", instruction.getMnemonicString());
					instructionData.put(
						"operands",
						instruction.getNumOperands() > 0 ? instruction.getDefaultOperandRepresentation(0) : ""
					);
					body.add(instructionData);
				}

				try {
					DecompileResults decompileResults = decompiler.decompileFunction(function, 30, monitor);
					if (decompileResults != null && decompileResults.decompileCompleted() && decompileResults.getDecompiledFunction() != null) {
						String source = decompileResults.getDecompiledFunction().getC();
						functionData.put("source", source != null ? source : "");
						functionData.put("decompiled", source != null ? source : "");
					}
					else if (decompileResults != null) {
						String errorMessage = decompileResults.getErrorMessage();
						functionData.put("decompile_error", errorMessage != null ? errorMessage : "Unknown decompilation error");
					}
				}
				catch (Exception exception) {
					functionData.put("decompile_error", exception.toString());
				}

				functions.add(functionData);
			}

			functions.sort(
				(left, right) -> Integer.compare(
					String.valueOf(right.get("source")).length(),
					String.valueOf(left.get("source")).length()
				)
			);

			DataIterator definedData = listing.getDefinedData(true);
			while (definedData.hasNext() && !monitor.isCancelled()) {
				Data data = definedData.next();
				try {
					if (data != null && data.hasStringValue()) {
						Map<String, Object> stringData = new LinkedHashMap<>();
						stringData.put("address", data.getAddress().toString());
						stringData.put("value", String.valueOf(data.getValue()));
						stringData.put("length", data.getLength());
						strings.add(stringData);
					}
				}
				catch (Exception exception) {
					// Ignore malformed string values and continue exporting other data.
				}
			}

			ExternalManager externalManager = program.getExternalManager();
			for (String libraryName : externalManager.getExternalLibraryNames()) {
				ExternalLocationIterator externalLocations = externalManager.getExternalLocations(libraryName);
				while (externalLocations.hasNext()) {
					ExternalLocation externalLocation = externalLocations.next();
					Map<String, Object> importData = new LinkedHashMap<>();
					importData.put("name", externalLocation.getLabel());
					importData.put("library", libraryName);
					importData.put(
						"address",
						externalLocation.getExternalSpaceAddress() != null
							? externalLocation.getExternalSpaceAddress().toString()
							: "external"
					);
					imports.add(importData);
				}
			}
		}
		finally {
			decompiler.dispose();
		}

		return result;
	}
}
